content/posts/replacing-the-redisgraph-parser.md

```diff
@@ -0,0 +1,109 @@
+
+> In the beginning the Universe was created.This has made a lot of people very angry and been widely regarded as a bad move.
+
+[59 commits, ~20,000 modified lines](https://github.com/RedisGraph/RedisGraph/pull/488), and a few months later, we've paid back a solid amount of technical debt and learned a few things along the way! RedisGraph queries are fully converted to ASTs by [libcypher-parser](https://github.com/cleishm/libcypher-parser), an Apache-licensed project authored and maintained over the past 4 years by Chris Leishman (on [Twitter](https://twitter.com/cleishm) and [Github](https://github.com/cleishm/) as cleishm).
+
+## When to repay technical debt
+
+> The extra effort that it takes to add new features is the interest paid on the debt.
+
+To use [Martin Fowler's language of technical debt](https://martinfowler.com/bliki/TechnicalDebt.html), our self-maintained grammar charged us interest every time we introduced or modified a syntactical construct. Gradual improvement was not an option, however, so we were reluctant to undertake the effort of a full replacement. We finally relented when working on the WITH clause, which is how Cypher allows for multiple queries to be chained into a single atomic execution. Unlike subqueries in SQL, which create a tree of independent executions, the WITH clause flattens query parts into a linear composition:
+
+![Multi-Part Query Railroad Diagram](/multipart_query.png)
+[Cypher railroad diagram for multi-part queries](https://s3.amazonaws.com/artifacts.opencypher.org/railroad/MultiPartQuery.html)
+
+Our original grammar was effective at interpreting a sequence of clauses as a self-contained query, but it would have required a major rewrite to interpret that same sequence as one scope capable of reading from and projecting into others. libcypher-parser had already solved this problem and many more, so we started the even more ambitious rewrite of scrapping our parser and building off an open-source project.
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
+
+RedisGraph is a better project for the issues opened, features requested, and contributions made by our open-source users. While integrating libcypher-parser, we've contributed code and feedback upstream, and collaboration that improves both projects is ongoing.
+
+The lesson that a vibrant open-source community enriches us all is not a new one, but I always find it heartening to be reminded of it!
```

Comment:
Change to: We all know that a vibrant open source community can enrich us, but I always find it heartening to be reminded.